---
title: "Assignment 05: Data wrangling"
format: 
  html:
    embed-resources: true
editor: source
---

# Instructions

Create a Quarto markdown document and answer the questions below using code blocks that generate the correct outputs. We encourage you to include explanatory text in your markdown document, however **each of your solutions should show how to generate the answer using R code**.

Write "robust" solutions wherever possible. A good rule of thumb for judging whether your solution is appropriately "robust" is to ask yourself "If I added additional observations or variables to this data set, or if the order of variables changed, would my code still compute the right solution?"

Make sure your markdown is nicely formatted -- use headers, bullets, numbering, etc so that the structure of the document is clear.

Make sure to "Render" your document before submission to confirm that all code blocks and formatting is as you expect.

When completed, title your markdown file as follows (replace `XX` with the assignemnt number, e.g. `01`, `02`, etc):

-   `netid-assignment_XX-Fall2024.qmd`

Submit both your markdown file (`.qmd`) and the generated HTML document (`.html`) on your Github site.



### Problems

### Wrangling a messy data set

Gasch AP, Spellman PT, Kao CM, Carmel-Harel O, Eisen MB, Storz G, Botstein D, Brown PO. Genomic expression programs in the response of yeast cells to environmental changes. Mol Biol Cell. 2000 Dec;11(12):4241-57. doi: 10.1091/mbc.11.12.4241. PMID: 11102521; PMCID: PMC15070.

Gasch et al. (2000) is another "classic" gene expression data set from the early days of functional genomics. The file [`gasch1k.csv`](https://github.com/Bio724D/Bio724D_2024_2025/raw/refs/heads/main/data/gasch1k.csv) is a subset of the data from that study.

A few notes about this data set:

* The first column of data describes the experimental conditions but this column has no header

* Subsequent columns are genes, with systematic gene IDs as headers

* The experimental conditions in the first column are not particularly consisten in terms of how they designate the conditions (and time points where appropriate)

#### Working with the Gasch data

Wrangle the data by carrying out the following steps, in order:

1. Import the data, fixing any column naming issues

2. Remove any columns (genes) that have more than 10 missing values ('NA')

3. Subset the rows of the data set to include ONLY the experimental conditions that have time points on the scales of minutes (conditions with "min", "minutes", etc).

4. Apply pivoting to create a "long" version of the data frame with "gene" and "expression" as their own columns.

5. CHALLENGING/ADVANCED: See if you can come up with a strategy for extracting the time information from the experimental condition column. The inconsistency of the naming scheme makes this a difficult and frustrating task.  A verbal description of what steps you'd take to solve this problem is sufficient. If you manage to write code to solve this, your string wrangling skills are already quite strong!











## Data lunch assignment

Identify something that you learned from the presentation or discussion on Thursday that you found valuable. Provide a brief reflection here (1-5 sentences) and include code or pseudo-code if useful. Hint: if you have specific code examples, consider adding them to your notebook as well.


